### PR TITLE
fix csv-export and payments chart

### PR DESCRIPTION
**Pull request by jackstar12**
Related: https://github.com/lnbits/lnbits/issues/1754

Adds an endpoint for retrieving payments in a csv format.
The provided colums are the same as previously exported in the frontend.

The `StreamingResponse` is used to avoid having to save the entire file locally, which results in better performance.

I also extracted the common logic of updating pending payments into its own function.
`check_payment_status` was replaced with `payment.check_status` because the first one calls the latter anyways while doing an additional query.

### CODE REVIEW

**Comment by jackstar12** [comment]
added the new payments chart here aswell. wanted to do a seperate pr but the new `/payments/history` endpoint uses the new function for updating pending payments aswell. This endpoint essentially returns the data as expected by the chart and makes it possible to reuse it in the future (maybe for a graph of the total node activity)

**Comment by jackstar12** [comment]
Note that I now removed the code I was already suspicious of in https://github.com/lnbits/lnbits/pull/1697 because it was now causing bugs when converting postgres dates
https://github.com/lnbits/lnbits/blob/95071eb87e94481a0b2e5bcd29517eaa1f8c8dee/lnbits/db.py#L50-L58

File: lnbits/db.py [hunk 1]
```diff
@@ -126,6 +120,24 @@ def timestamp_placeholder(cls):
         else:
             return "?"
 
+    @classmethod
+    def truncate_date(cls, col: str, trunc: DateTrunc) -> str:
```
Comment (motorina0) [inline lnbits/db.py:124]: - value format/truncate should take place at the Business Logic layer (python) and not at the Persistence layer (SQL)

File: lnbits/core/crud.py [hunk 2]
```diff
@@ -619,6 +627,64 @@ async def update_payment_extra(
...
+            balance = wallet.balance_msat
+        else:
+            raise ValueError("Unknown wallet")
+    else:
+        balance = await get_total_balance()
+
+    # since we dont know the balance at the starting point, we take the current balance and walk backwards
+    results: list[PaymentHistoryPoint] = []
+    for row in transactions:
+        print(row)
```
Comment (motorina0) [inline lnbits/core/crud.py:677]: - remove `print()`

File: lnbits/db.py [hunk 3]
```diff
@@ -126,6 +120,24 @@ def timestamp_placeholder(cls):
...
+            formats = {
+                "hour": "%Y-%m-%d %H:00:00",
+                "day": "%Y-%m-%d 00:00:00",
+                "month": "%Y-%m-01 00:00:00",
+            }
+            if trunc in formats:
+                return f"strftime('{formats[trunc]}', {col}, 'unixepoch')"
+        elif trunc in ("day", "hour", "month"):
+            return f"date_trunc('{trunc}', {col})"
+        raise ValueError(f"Invalid truncation: {trunc}")
```
Comment (callebtc) [inline lnbits/db.py:139]: What exactly does this do? In sqlite, we used to store unix timestamps whereas in postgres, we used the `date` type. Does this affect anything? Would be nice to clarify.

File: lnbits/core/crud.py [hunk 4]
```diff
@@ -619,6 +627,63 @@ async def update_payment_extra(
...
+        filters = Filters()
+    where = ["(pending = False OR amount < 0)"]
+    values = []
+    if wallet_id:
+        where.append("wallet = ?")
+        values.append(wallet_id)
+
+    transactions = await db.fetchall(
+        f"""
+        SELECT {db.truncate_date('time', group)} date,
```
Comment (callebtc) [inline lnbits/core/crud.py:655]: An equivalent of `db.truncate_date` should be applied in the frontend (chart) that needs the truncated time stamps, not in the db level.

File: lnbits/db.py (hunk 1 above)
Comment (callebtc) [reply lnbits/db.py:124]: Agreed. There are three options to truncate:

- in the db
- in the endpoint
- in the frontend

only the fronend needs it (for this specific use case of charting the payments), so it should be handled there.

**Comment by callebtc** [comment]
There are multiple undocumented logic changes in this PR, like `update_pending_payments` and returned `get_payments` calls. Please clarify each change so we can properly review it! Thanks 🙏

File: lnbits/db.py (hunk 3 above)
Comment (jackstar12) [reply lnbits/db.py:139]: No, it just affects the way those timestamps are being used in the query.

**Comment by jackstar12** [comment]
> There are multiple undocumented logic changes in this PR, like `update_pending_payments` and returned `get_payments` calls. Please clarify each change so we can properly review it! Thanks 🙏

From above (regarding `update_pending_payments`):
I also extracted the common logic of updating pending payments into its own function.
check_payment_status was replaced with payment.check_status because the first one calls the latter anyways while doing an additional query.

Which `get_payments` calls do you mean?

**Comment by talvasconcelos** [comment]
Fixes #1754

**Comment by jackstar12** [comment]
@callebtc @motorina0 
what about a middleground where the csv-export just fetches all payments into the frontend and creates the csv there? Altough less efficient it would make for less added code. Except retrieving csv through api is desired.

**Comment by jackstar12** [comment]
I think the balance history should be on the server tough for above stated reasons (and for creating a admin balance chart of the entire instance)

**Pull request closed without merging**